cmd/container-disk-v2alpha/main.c

```diff
@@ -0,0 +1,142 @@
+    copy_path = (char *)arg;
+
+    for (;;) {
+        sleep(1);
+        if (stat(copy_path, &st) == -1) {
+            printf("socket %s does not exist anymore\n", copy_path);
+            exit(0);
+        }
+    }
+}
+
+int main(int argc, char **argv) {
+    char *copy_path = NULL;
+    char *copy_path_dir;
+    bool health_check = false;
+
+    int c;
+    while(1) {
+        static struct option long_options[] = {
+            /* These options don’t set a flag.
+                We distinguish them by their indices. */
+            {"copy-path",    required_argument, 0, 'c'},
+            {"health-check", no_argument,       0, 'p'},
+            {0, 0, 0, 0}
+        };
+        /* getopt_long stores the option index here. */
+        int option_index = 0;
+        c = getopt_long(argc, argv, "c:p", long_options, &option_index);
+
+        /* Detect the end of the options. */
+        if (c == -1) {
+            break;
+        }
+
+        switch (c) {
+            case 'c':
+                copy_path = optarg;
+                copy_path_dir = strdup(copy_path);
+                dirname(copy_path_dir);
+                break;
+
+            case 'p':
+                health_check = true;
+                break;
+
+            case '?':
+                exit(1);
+            default:
+                abort();
+        }
+    }
+
+    struct stat st = {0};
+    if (health_check) {
+        if (stat(READINESS_PROBE_FILE, &st) == -1) {
+            printf("readiness probe %s does not exist\n", READINESS_PROBE_FILE);
+            exit(1);
+        } else {
+            exit(0);
+        }
+    }
+
+    if (stat(copy_path_dir, &st) == -1) {
+        if (mkdir(copy_path_dir, 0777) != 0) {
+            printf("failed to create disk directory %s\n", copy_path_dir);
+            exit(1);
+        }
+    }
+
+    struct sockaddr_un address;
+    /*
+    * For portability clear the whole structure, since some
+    * implementations have additional (nonstandard) fields in
+    * the structure.
+    */
+    memset(&address, 0, sizeof(struct sockaddr_un));
+    address.sun_family = AF_UNIX;
+    strcat(copy_path, ".sock");
+    strcpy(address.sun_path, copy_path);
+
+    int fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (fd == -1) {
+        printf("failed to create socket on %s\n", copy_path);
+        exit(1);
+    }
+
+    if (bind(fd, (struct sockaddr*)(&address), sizeof(struct sockaddr_un)) == -1) {
+        printf("failed to bind socket %s\n", copy_path);
+        exit(1);
+    }
+
+    if (listen(fd, LISTEN_BACKLOG) == -1) {
+        printf("failed to listen socket %s\n", copy_path);
+        exit(1);
+    }
+
+    // Create readiness probe
+    FILE *probe;
+    probe = fopen(READINESS_PROBE_FILE, "w");
+    if (probe == NULL) {
+        printf("failed to create readiness probe\n");
+        exit(1);
+    }
+    fclose(probe);
+
+    pthread_t socket_check_thread;
+    pthread_create(&socket_check_thread, NULL, socket_check, (void *)copy_path);
```

Comment:
I think that you just forgot to remove the `pthread` related code.

Reply:
not really, in the end, we need it, `accept` block the code until it receives any requests, and socket remove does not unblock it, so we need socket check to exit from the main function and release all resources.

Reply:
This feels like an anti-pattern. Can't we just `listen` on the socket, then use `select` to block until a connection is made or the socket is deleted? It would remove the need for the socket_check_thread entirely I think.

Reply:
I do not sure I am following you, I can see that `select` can wait for number of FD to be ready with some specified timeout, in our case we have `accept` that blocked until it receives any request and the tread is checking the socket file, to exit the main when the virt-handler will clean the socket. Maybe do you have some examples of select usage that will apply to our use case?

Reply:
On line 131, instead of blindly doing an `accept`, which will block. do a `select`--which can also block indefinitely. When select returns, one of two things is true:

1) the socket has been closed, and exit(1) is the next step
2) an incoming connection has been initiated. `accept` is guaranteed not to block.

Thus "select -> check -> accept" alleviates the need for a separate thread that simply monitors the health of the socket. the main thread will be woken up immediately if it becomes unhealthy.